esphome/components/lc709203f/lc709203f.cpp

```diff
@@ -0,0 +1,302 @@
+        snprintf(this->error_code_buffer_, 50, "CRC error reading from register 0x%02X", register_to_read);
+        this->status_set_warning(this->error_code_buffer_);
+      } else {
+        *register_value = ((uint16_t) read_buffer[4] << 8) | (uint16_t) read_buffer[3];
+        return i2c::NO_ERROR;
+      }
+    }
+  }
+
+  // If we get here, we tried LC709203F_I2C_RETRY_COUNT times to read the register and
+  //  failed each time. Set the register value to 0 and return the I2C error code or 0xFF
+  //  to indicate a CRC failure. It will be up to the higher level code what to do when
+  //  this happens.
+  *register_value = 0x0000;
+  if (return_code != i2c::NO_ERROR) {
+    return return_code;
+  } else {
+    return 0xFF;
+  }
+}
+
+uint8_t Lc709203f::set_register_(uint8_t register_to_set, uint16_t value_to_set) {
+  i2c::ErrorCode return_code;
+  uint8_t write_buffer[5];
+
+  // Note: We don't actually send byte[0] of the buffer. We include it because it is
+  //  part of the CRC calculation.
+  write_buffer[0] = (this->address_) << 1;
+  write_buffer[1] = register_to_set;
+  write_buffer[2] = value_to_set & 0xFF;         // Low byte
+  write_buffer[3] = (value_to_set >> 8) & 0xFF;  // High byte
+  write_buffer[4] = this->crc8_(write_buffer, 4);
+
+  for (int i = 0; i <= LC709203F_I2C_RETRY_COUNT; i++) {
+    // Note: we don't write the first byte of the write buffer to the device.
+    //  This is done automatically by the write() function.
+    return_code = this->write(&write_buffer[1], 4, true);
+    if (return_code == i2c::NO_ERROR) {
+      return return_code;
+    } else {
+      snprintf(this->error_code_buffer_, 50, "Error code %d when writing to register 0x%02X", return_code,
+               register_to_set);
+      this->status_set_warning(this->error_code_buffer_);
+    }
+  }
+
+  // If we get here, we tried to send the data LC709203F_I2C_RETRY_COUNT times and failed.
+  //  We return the I2C error code, it is up to the higher level code what to do about it.
+  return return_code;
+}
+
+uint8_t Lc709203f::crc8_(uint8_t *byte_buffer, uint8_t length_of_crc) {
+  uint8_t crc = 0x00;
+  const uint8_t polynomial(0x07);
+
+  for (int j = length_of_crc; j; --j) {
+    crc ^= *byte_buffer++;
+
+    for (int i = 8; i; --i) {
+      crc = (crc & 0x80) ? (crc << 1) ^ polynomial : (crc << 1);
+    }
+  }
+  return crc;
+}
+
+void Lc709203f::set_pack_size(uint16_t pack_size) {
+  static const uint16_t PACK_SIZE_ARRAY[6] = {100, 200, 500, 1000, 2000, 3000};
+  static const uint16_t APA_ARRAY[6] = {0x08, 0x0B, 0x10, 0x19, 0x2D, 0x36};
+  float slope;
+  float intercept;
+
+  this->pack_size_ = pack_size;  // Pack size in mAH
+
+  // The size is used to calculate the 'Adjustment Pack Application' number.
+  // Here we assume a type 01 or type 03 battery and do a linear curve fit to find the APA.
+  for (int i = 0; i < sizeof(PACK_SIZE_ARRAY) / sizeof(PACK_SIZE_ARRAY[0]); i++) {
```

Comment:
I think you can use `uint8_t` here, too, since, if I'm understanding this correctly, you're basically just doing `for (uint8_t i = 0; i < 6; i++)`...
```suggestion
  for (uint8_t i = 0; i < sizeof(PACK_SIZE_ARRAY) / sizeof(PACK_SIZE_ARRAY[0]); i++) {
```